### ISSUE TITLE
[BUG] pennylane.numpy tensors become trainable multiplied by numpy.ndarray

### ISSUE THREAD

**Issue by timmysilv** [PennyLaneAI/pennylane#4350]
### Expected behavior

The returned tensor should also have `requires_grad=False`

### Actual behavior

The returned tensor has `requires_grad=True`

### Additional information

This is troublesome because it happens in things like `qml.transforms.one_qubit_decomposition`. For example,
```pycon
>>> U = qnp.array([[0, 1], [1, 0]], requires_grad=False)
>>> qml.QubitUnitary(U, 0).decomposition()
[RZ(tensor(1.57079633, requires_grad=True), wires=[0]),
 RY(tensor(3.14159265, requires_grad=True), wires=[0]),
 RZ(tensor(10.99557429, requires_grad=True), wires=[0])]
```
`requires_grad` is suddenly True in the results! [This line in _convert_to_su2](https://github.com/PennyLaneAI/pennylane/blob/b781be05d1b21bcb8e874ce8d13f03ba283270f4/pennylane/transforms/decompositions/single_qubit_unitary.py#L44) is the culprit. The first term (the `cast_like`) still has `requires_grad=False` set, but once multiplied with the second term (the `exp`, a `numpy.ndarray`), the resulting value has `requires_grad=True`.

#### Note for fixer

We are not setting trainable parameters in `validate_and_expand_adjoint` (see #4365) partially because of this bug. While fixing this, consider adding it to that function and adjusting tests accordingly. Deciding what to do with `tape.trainable_params` is always messy, so I'll let you decide :)

### Source code

```shell
from pennylane import numpy as qnp
import numpy as np
qnp.array(0.1, requires_grad=False) * np.array(0.2)
```

### System information

```shell
pl dev
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by timmysilv** [comment]
Here's a summary of findings so far. Consider this example:
```python
dev = qml.devices.experimental.DefaultQubit2()
mat = qml.RZ.compute_matrix(2.2)
x = qnp.array(0.1, requires_grad=True)
U = qnp.array(mat, requires_grad=False)

config = qml.devices.experimental.ExecutionConfig(gradient_method="adjoint")
qs = qml.tape.QuantumScript([qml.RX(x, 0), qml.QubitUnitary(U, 0)], [qml.expval(qml.PauliZ(0))])

tapes, fn, config = dev.preprocess([qs], execution_config=config)
dev.compute_derivatives(tapes, execution_config=config)
# ((array(-0.09983342), array(0.), array(-0.08897228), array(-2.16840434e-19)),)
```
Unfortunately, this computes the derivate of each of the four parameters (QubitUnitary's ZYZ decomposition makes it have 3 parameters - you can print `tapes[0].data` here if you're curious). Similarly, consider this equivalent circuit and gradient computation:
```python
@qml.qnode(dev, diff_method="adjoint")
def circuit(x, U):
    qml.RX(x, 0)
    qml.QubitUnitary(U, 0)
    return qml.expval(qml.PauliZ(0))

qml.grad(circuit)(x, U)
# -0.09983341664682814
```
as expected, you only get the one gradient result! However, the gradient of all 4 parameters is still computed internally, meaning we are doing unnecessary computation. But why don't we get all 4 gradients returned? Or the gradient of `U` perhaps? It's because `qml.grad` converts our parameters from `qml.numpy.tensor` objects to autograd `ArrayBox` instances _before_ we decompose our circuit. So in this case, our autograd interface code sees something like `[<autograd.numpy.numpy_boxes.ArrayBox object at 0x152891940>, tensor(1.1, requires_grad=True), tensor(0., requires_grad=True), tensor(1.1, requires_grad=True)]`. Autograd itself does not care about `requires_grad`; it only cares about whether you are an `ArrayBox` or not. So `qml.grad` only returns the gradient for the first parameter.

If we instead change the default s.t. multiplying a `tensor` with an `ndarray` preserves the `requires_grad` value, other things break, like this circuit:
```python
dev = qml.devices.experimental.DefaultQubit2()

@qml.qnode(dev)
def circuit(x, coeff):
    qml.RX(x, 0)
    return qml.expval(qml.Hamiltonian([coeff], [qml.PauliZ(0)]))

x = qnp.array(0.1, requires_grad=True)
coeff = qnp.array(1.1, requires_grad=False)

qml.grad(circuit)(x, coeff)
# NonDifferentiableError: 1.0945045818058285 is non-differentiable. Set the requires_grad attribute to True.
```
Note that this only breaks with the new device interface because of the multiplication in `measure.sum_of_terms_method`. Perhaps we should make changes there as well, I'll look into it. But one option is to update this error to ask users to use vanilla numpy instead if they don't want trainability in these strange cases.

**Comment by timmysilv** [comment]
closing as wont-fix. the current behaviour of things is better than the alternative, so we'll have this for documentation's sake.

tl;dr - the unexpected conversion of `requires_grad` from False to True during evaluation does _not_ alter results, but adds an unnecessary computational overhead. if you want to avoid this overhead, define the non-trainable parameters with vanilla numpy (`import numpy as np`) instead of pennylane's numpy (`from pennylane import numpy as np`, with `requires_grad` set to False)
